### ISSUE TITLE
Confusion on many_to_many (uuid::bytea coercion)

### ISSUE THREAD

**Issue by ryanwinchester** [bitwalker/uniq#10]
I have 

```elixir
config :my_app, MyApp.Repo,
  migration_primary_key: [type: :binary_id],
  migration_foreign_key: [type: :binary_id]
```
(which sets it to `uuid` type in DB)

```elixir
  @primary_key {:id, Uniq.UUID, version: 7, autogenerate: true}
  @foreign_key_type Uniq.UUID
      
  schema "users" do
    # ...
    many_to_many :locations, Location, join_through: "user_locations"
  end
  ```
  
  And if I call `Repo.preload(user, [:locations])` I get the following error:
  
 ```
 [error] Task #PID<0.730.0> started from #PID<0.723.0> terminating
** (Postgrex.Error) ERROR 42846 (cannot_coerce) cannot cast type uuid to bytea

    query: SELECT l0."id", l0."email", l0."name", l0."inserted_at", l0."updated_at", u1."user_id"::bytea
            FROM "locations" AS l0 INNER JOIN "user_locations" AS u1 ON l0."id" = u1."location_id"
            WHERE (u1."user_id" = ANY($1)) ORDER BY u1."user_id"::bytea

    (ecto_sql 3.9.1) lib/ecto/adapters/sql.ex:913: Ecto.Adapters.SQL.raise_sql_call_error/1
    (ecto_sql 3.9.1) lib/ecto/adapters/sql.ex:828: Ecto.Adapters.SQL.execute/6
    (ecto 3.9.2) lib/ecto/repo/queryable.ex:229: Ecto.Repo.Queryable.execute/4
    (ecto 3.9.2) lib/ecto/repo/queryable.ex:19: Ecto.Repo.Queryable.all/3
    (ecto 3.9.2) lib/ecto/repo/preloader.ex:272: Ecto.Repo.Preloader.fetch_query/8
    (elixir 1.14.2) lib/task/supervised.ex:89: Task.Supervised.invoke_mfa/2
    (elixir 1.14.2) lib/task/supervised.ex:34: Task.Supervised.reply/4
    (stdlib 4.1.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Function: &:erlang.apply/2
    Args: [#Function<8.114411606/1 in Ecto.Repo.Preloader.maybe_pmap/3>,
          [#Function<20.114411606/1 in Ecto.Repo.Preloader.prepare_queries/6>]]
```

What is the best way to make this work?

**Comment by cospin** [comment]
I just ran into the same problem.

**Comment by ryanwinchester** [comment]
I've worked around it for now by doing this, unfortunately.

```elixir
  schema "users" do
    # ...
    # many_to_many :locations, Location, join_through: "user_locations"
    has_many :user_locations, UserLocation
    has_many :locations, through: [:user_locations, :location]
  end
  ```

**Comment by bitwalker** [comment]
I believe the issue is that the `Ecto.ParameterizedType` implementation defaults to a type of `:binary`, whereas the underlying column type is `uuid`, hence the cast to `bytea` in the query. Right now there isn't a way to override that to be `:uuid`, but that's easy enough to add, since we already support switching between `:binary` and `:string` depending on the UUID representation chosen, by providing the `:dump` option with a format other than `:raw`.

Unfortunately there isn't a way (AFAIK) to automatically set the type based on the underlying schema, and furthermore, I'm not sure if there is any validation done by the database on UUID values to ensure they are of a specific version/format, if so, we'd have to restrict which types of UUID formats can be used with the `:uuid` type. The built-in `Ecto.UUID` type for example, only supports generating v4 UUIDs.

In the near term, I'm tempted to just allow overriding the type returned by `type/1` with `:uuid`, but I'd like to be smarter about it if possible.

**Comment by bitwalker** [comment]
@ryanwinchester @cospin I've pushed a patch to main which allows you to specify a `type: :uuid | :binary | :string` value when defining the `@primary_key` attribute, e.g. ` @primary_key {:id, Uniq.UUID, version: 7, autogenerate: true, type: :uuid}`.

Can one of you test whether or not that actually works? My suspicion is that the database might validate the UUID metadata (i.e. what version of UUID it is), but if not, then that makes for an easy fix.

I believe I chose to default to `:binary` for perf reasons, since it avoids the unnecessary encoding/decoding step, and is also how you benefit from the lexicographical ordering characteristics (IIRC); but I might make the default type `:uuid` rather than `:string` when `:dump` is not `:raw`, since that likely aligns better with how the data is stored in that case.

**Comment by cospin** [comment]
I just tested and it works without issues now! Thank you!

**Comment by ryanwinchester** [comment]
@bitwalker also tested main on my use-case and it works now 🙏 👍

**Comment by ryanwinchester** [comment]
Also `uuid` column in postgres can be any type of UUID, but must be a valid UUID syntax but it doesn't care which version or validate the version. Also UUID v6/v7 are ordered as expected.

**Comment by ryanwinchester** [comment]
Fixed in 3df2ec076756a47926e48aec175f36b065b0c8a2
